ENH: Support for Multi-Wavelength (>2) NIRS/SNIRF Data Processing

zEdS15B3GCwq opened the pull request:
#### Reference issue (if any)
Started discussion in #13401. Most of the description is copied from there with changes.

The same idea has also been raised in issue https://github.com/mne-tools/mne-python/issues/9816, but that discussion did not result in a PR. In short, (some) Shimadzu fNIRS devices work with 3 wavelengths, and in the future there might be other devices with more, but mne's SNIRF I/O and processing functions are coded to accept only 2 wavelengths.

From my perspective, these changes would be mostly beneficial as I'm using a Shimadzu device. While the device doesn't support .snirf output, there are ways to convert the data to that format. And while it's possible to preprocess and convert to Hb data first in some other tool(s) and then import to mne, it would be ideal to keep the entire processing pipeline to one tool.

### Describe the new feature or enhancement

Goal: extend MNE-Python's NIRS processing to support an arbitrary number of wavelengths (≥2) instead of being limited (hard-coded in several places) to two.

I'd like to discuss implementation details and to explore devs' opinion on how feasible such changes would be. For this reason, this a _draft PR_, and some important points (like tests) haven't been addressed (at all) yet.

### Describe your proposed implementation

I've done some exploration in the codebase to identify relevant bits of code, both manually and with the help of Claude 4. Based on this, it seems that relevant parts of the code are in SNIRF I/O (mne.io.snirf) and NIRS preprocessing (mne.preprocessing.nirs).

### Major changes
- **Channel data structure**: with 2 wavelengths, channel data is stored as ABAB... . All functions handling OD and amplitude data will need to be updated not to rely on a fixed [::2], [1::2] sequence. I listed this as a major change due to the difficulty of locating every function that works this way. Claude thinks it's been able to find all of them (see list of files to be changed below). It still requires devs' experience to know if there are some I might not have been able to find.
- **Beer-Lambert calculation (algorithmic change)**: The current implementation uses pseudo-inverse `pinv()` to calculate `ΔOD = E × L × PPF × Δc`. I think it's possible to update this with minimal changes, just changing the EL matrix to n x 2 where n is the number of wavelengths, instead of the current 2 x 2. After Beer-Lambert, excess channels will need to be dropped.
- **Scalp coupling index (algorithmic change)**: I'm not aware of any reports where SCI was defined for >2 wavelengths. My idea here is to calculate all possible pairs of correlation, and use the minimum value. Intuitively, this should be correct as all wavelength channels should be correlated.

### Minor changes
- **SNIRF file reading**: checks and validation steps need to be updated. While there's a number of validation checks, the overall complexity of the code changes isn't high.

### List of affected files and functions:
- [/mne/io/snirf/_snirf.py: `class RawSNIRF(BaseRaw)`](https://github.com/mne-tools/mne-python/blob/main/mne/io/snirf/_snirf.py#L151) has a simple check for 2 wavelengths and an error message.
- [/mne/preprocessing/nirs/_beer_lambert_law.py: `beer_lambert_law()`](https://github.com/mne-tools/mne-python/blob/main/mne/preprocessing/nirs/_beer_lambert_law.py#L17) logic needs to be updated to >=2 wavelengths, plus it will need to remove excess channels after assigning HbO/R data to the first 2.
- [/mne/preprocessing/nirs/_beer_lambert_law.py: `_load_absorption()`](https://github.com/mne-tools/mne-python/blob/main/mne/preprocessing/nirs/_beer_lambert_law.py#L88) is hard-coded to 2 wavelengths.
- [/mne/preprocessing/nirs/_scalp_coupling_index.py: `scalp_coupling_index()`](https://github.com/mne-tools/mne-python/blob/main/mne/preprocessing/nirs/_scalp_coupling_index.py#L13) needs an algorithmic update
- [/mne/preprocessing/nirs/nirs.py: `_check_channels_ordered()`](/mne/preprocessing/nirs/nirs.py#L87) has various validation checks and [`_fnirs_spread_bads()`](/mne/preprocessing/nirs/nirs.py#L272) needs to spread bad markings over all wavelengths in group

I'm hoping to receive feedback on the correctness of the above proposed algorithmic changes. For n=2 wavelengths, the new calculation methods wouldn't introduce any new changes (E in Beer-Lambert would still be 2x2, and there is only one correlation pair so the minimum is irrelevant), and if necessary, the current implementation could be preserved in a separate code path. Current tests for n=2 should also work.

### Describe possible alternatives

The algorithmic changes (Beer-Lambert and SCI) could be implemented in another way. The Beer-Lambert pseudo-inverse should work with n>2 with minimal changes and the change seems mathematically correct. The SCI calculation using the minumum of all correlations appers to be the most conservative in terms of being able to spot any coupling issues, but without studies to rely on, it's a choice that should be discussed.

### Additional context
I'll open another PR for the `mne-nirs` project once this PR has advanced a bit. In particular, the scalp coupling algorithm should be agreed on, since that project has a very similar, windowed implementation of it that probably should be updated with the changes here in mind.

As I mentioned above, the current code is untested, as in I haven't actually run it with any .snirf files yet. I'm sorry for this, I'll try to do some testing as soon as possible.

larsoner reviewed (commented):
Looks like a reasonable start to me!

larsoner commented:
Looks like style CIs are unhappy -- until they are, most tests won't run

zEdS15B3GCwq commented on mne/preprocessing/nirs/_scalp_coupling_index.py:
```diff
@@ -56,14 +56,52 @@ def scalp_coupling_index(
+    n_wavelengths = len(set(ch_wavelengths))
+
+    # freqs = np.array([raw.info["chs"][pick]["loc"][9] for pick in picks], float)
+    # n_wavelengths = len(set(unique_freqs))
```
Left it there to remind me of something, forgot to remove apparently.

zEdS15B3GCwq commented:
I know why the `mne/io/hitachi/tests/test_hitachi.py::test_hitachi_basic` tests are failing, but I need some feedback on how to best fix it.

### Multiple actual wavelengths stored in `info`

The Hitachi tests use data that is special because it has both overall _nominal_ (`Wave[nm]`) and what I assume to be per-channel, _actual_ wavelength values (`Wave Length`):

```csv
Wave[nm],695,830
Wave Length,CH1(703.6),CH1(829.0),CH2(703.9),CH2(829.3),...,CH52(704.4),CH52(829.0)
```

The nominal values appear in channel names (e.g. "S1_D1 695", "S1_D1 830"), while the actual ones are stored in the _info_ structure (`[raw.info["chs"][pick]["loc"][9]`). There are 2 nominal wavelengths, satisfying the requirement that there can only be 2, but basically each channel has its own actual wavelength value (some are equal, some not). This means that, in a way, MNE has been "multi-wavelength" all along by allowing any wavelength value in the _info_ fields.

At first look, this seems to be acceptable, as validation checks do not explicitly test the actual values in any way except that it cannot contain _NaN_. All the rest of the checks are performed on `pair_vals`, which contain the nominal wavelength pair.

Validation only checks if wavelengths in _info_ are Nan:
https://github.com/mne-tools/mne-python/blob/0a06818fc780ce2a344b7fe70c7acdb8ac90ca9f/mne/preprocessing/nirs/nirs.py#L157-L163

### MNE inconsistent in what is stored in `info`

On the other hand, there are many indications in the code that this is not the expected behaviour, in fact, **it is incorrect**:

* To my understanding, MNE doesn't have a clear way of storing nominal and actual wavelengths. Keeping one in the channel name, and the other in `info` without excplicit separation of their purpose seems to be a hack at best. Perhaps I misunderstand the intent behind MNE's current handling of the data, however...
* Actual wavelength values are also supported by SNIRF, but the SNIRF reader does not even check for `wavelengthActual` fields and only uses `wavelengthIndex` values which select one of the nominal values. This is enforces that wavelength values stored in `info` and channel names are the same (except that the names are rounded to integers). So there's the confusion that in one format the `info` structure has nominal, in another it has actual values. This cannot/should not be the intention of MNE's design... or, well, we could say that the `info` has the _most accurate available_ wavelength value.
* A test in [mne.preprocessing.nirs.tests.test_beer_lambert_law.py](https://github.com/mne-tools/mne-python/blob/0a06818fc780ce2a344b7fe70c7acdb8ac90ca9f/mne/preprocessing/nirs/tests/test_beer_lambert_law.py#L58) explicitly mentions that it's an error "if channel naming frequency doesn't match what is stored in loc[9]". The test didn't actually work as intended as there was no validation check to test this. I added a new validation check which corrects this, but causes the Hitachi code to fail.
* The most important place where the wavelengths matter is in the Beer-Lambert Law calculation. Assuming that the actual wavelengths can be different for each channel (as they do in the Hitachi data), one should either use the nominal wavelengths to calculate overall absorption coefficients, or get individual coeffs for each channel based on the actual wavelengths. The code does neither, it **incorrectly calculates absorption coefficients for _all_ channels based _only on the 1st pair_ of actual wavelengths**. Not the nominal ones. This must be a mistake, as actual wavelengths can deviate from the expected value, and if the first pair deviates significantly, then calculation for all the channels will be off. 

BLL calculation calculates absorption coefficients once, used for all channel pairs:
https://github.com/mne-tools/mne-python/blob/0a06818fc780ce2a344b7fe70c7acdb8ac90ca9f/mne/preprocessing/nirs/_beer_lambert_law.py#L48-L72

Absorption coefficient calculation only returns values for the 1st frequency pair:
https://github.com/mne-tools/mne-python/blob/0a06818fc780ce2a344b7fe70c7acdb8ac90ca9f/mne/preprocessing/nirs/_beer_lambert_law.py#L107-L112

### Tests fail because they expect same values in `info` and channel names

Finally, the tests are failing now because I assumed that wavelengths in the `info` field must be the same as in the channel names. I added a validation test that is based on that assumption:
https://github.com/zEdS15B3GCwq/mne-python/blob/926b064be5c564fd2fb489481908b1b1a6b223c4/mne/preprocessing/nirs/nirs.py#L167-L173
```python
        # test if the info structure has the same number of freqs as pair_vals
        if len(pair_vals) != len(set(all_freqs)):
            picks = _throw_or_return_empty(
                f"The {error_word} in info must match the number of wavelengths, "
                f"but the data contains {len(set(all_freqs))} wavelengths instead.",
                throw_errors,
            )
```

This code compares `pair_vals`, nominal wavelength array [calculated from channel names](https://github.com/mne-tools/mne-python/blob/0a06818fc780ce2a344b7fe70c7acdb8ac90ca9f/mne/preprocessing/nirs/nirs.py#L65), to wavelengths in the `info` structure. Obviously, this check is pointless if `info` can contain arbitrary wavelengths. It is also wrong, as `set()` or `np.uniqe` is guaranteed to cause a mess with such data.

### How to proceed (feedback please)

I can see several potential paths:

1. **Nominal and actual wavelengths have to be the same.** This clarifies that `info` stores nominal wavelengths. Hitachi reader has to be fixed to store nominal values in `info`, discard actual ones. The failing test will pass.
2. **Allow actual wavelengths in info, use nominal ones in BLL.** This keeps the current way of handling information, but fixes BLL. My new validation check will be removed, and the failing tests will pass. However, if actual wavelengths are not used anywhere, why bother with storing them in the first place?
3. **Allow actual wavelengths in info and use them in BLL.** The current way of storing wavelengths in `info` is preserved. BLL is fixed, and its accuracy is potentially improved in cases where actual wavelength information exists. In the implementation, nominal values will be used to define the size of channel groups, but for each channel the absorption coefficients will be based on their individual actual wavelength. The new validation check will be removed, the test will pass.

Even though I wrote that the new validation check can be removed, it might be useful to add _some_ kind of validation for the values stored in `info`. As of now, it's left to the respective format readers to decide what to put in it.

### Others

* Docs build locally without isses on Ubuntu 24LTS on WSL2. You mentioned that you're going to make CircleCI work, so I'm assuming the docs are not an issue.
* Both on `main` and in this PR branch, I have some failing tests when the suite is run on Windows. Is that a problem?
```bash
> uv run pytest -m "not ultraslowtest" mne
...
6164 passed 2 failed
FAILED mne/viz/_brain/tests/test_brain.py::test_brain_screenshot[pyvistaqt] - assert (450, 525, 3) == (np.int64(600...int64(700), 3)
FAILED mne/viz/tests/test_epochs.py::test_plot_epochs_clicks[qt] - KeyError: 'MEG 0113'
ERROR mne/viz/_brain/tests/test_brain.py::test_brain_time_viewer[pyvistaqt] - AssertionError: []
```
* When the tests are run on Ubuntu24LTS/WSL2, `mne/viz/tests/test_epochs.py` passes (failed on Window), but many tests in `mne/viz` fail, for example:
```bash
mne/viz/_brain/tests/test_brain.py X Error of failed request:  BadWindow (invalid Window parameter)
```

Do you have any suggestions how I should configure WSL2 for testing? I can run `xeyes` and `xcalc`, and several MNE tests successfully pop up windows and pass.

zEdS15B3GCwq commented:
@larsoner , as you requested, I've added a `13408.newfeature.rst`, but placed it under `doc/changes/dev/` as I saw other files there and the contributing guide also says so. Let me know if it's not the correct place. I've also added my name to `doc/changes/names.inc`, and flipped to 'ready for review'.

Would you like me to add some tests as well?

larsoner commented on doc/changes/dev/13408.newfeature.rst:
```diff
@@ -0,0 +1,2 @@
+Add support for multi-wavelength NIRS processing to :func:`mne.preprocessing.nirs.beer_lambert_law`, :func:`mne.preprocessing.nirs.scalp_coupling_index`, :mod:`mne.preprocessing.nirs.nirs`, and SNIRF reader :class:`mne.io.snirf._snirf.RawSNIRF`, by :newcontrib:`Tamas Fehervari`.
```
That submodule isn't publicly documented so we can't link to it. Better to just link to the user-facing stuff
```suggestion
Add support for multi-wavelength NIRS processing to :func:`mne.preprocessing.nirs.beer_lambert_law`, :func:`mne.preprocessing.nirs.scalp_coupling_index`, and SNIRF reader :class:`mne.io.snirf._snirf.RawSNIRF`, by :newcontrib:`Tamas Fehervari`.
```

larsoner commented:
Fixtures are good for iteration, and can work. But at the end of the day we need to be able to operate on end-user files that people will want to process, so it's nice to have a concrete example of a binary someone could collect and want to read+process with MNE-Python. And assuming it's a valid file and valid use case, it's also not the sort of thing we typically *change* -- we usually *add* files for additional use cases and systems while keeping the old ones around and continuing to test those. Files in `mne-testing-data` also have the advantage that anyone can easily access, download, process etc. manually outside a strict `pytest` context, which isn't true (at least not easily/trivially) of pytest-fixture-based tests. That said, I think a fixture-based test is okay to start, and we can always add a file obtained using whatever process an end-user would actually use (`labnirs2snirf`, or maybe someday acquisition systems support outputting `snirf` or some other format we can read) and process that, too.

zEdS15B3GCwq commented:
@larsoner, I can agree with most of your points about having files. My comment still stands, however, that any data I convert in my own way is probably not going to be representative of what end-users might actually use, as alternative conversion methods may have differences. 

I've just pushed a partial solution for multi-wavelength testing, mostly with the intention to ask you whether it's going in a good direction and because I'm a bit stuck with `mne/preprocessing/nirs/tests/test_nirs.py`.

- In this commit I used fixtures to generate the snirf / nirs data. In my defence, several existing tests seem to be doing the same, manually constructing a snirf file or raw data.
- Another point of note is that I didn't integrate with existing tests but added new tests instead. I've found that existing tests often use logic specific to 2 wavelengths or use fixed values found in a particular dataset. Mixing multi-wavelength checks in those tests would've required a lot of if-else branches.

The place I've hit a snag is in `mne/preprocessing/nirs/tests/test_nirs.py`. It has too many tests and I don't want to bloat the code too much by duplicating a number of these tests with the multi-wavelength data and some modifications. This is the point where I think having a test file as a starting point would make things easier. Mixing fixture with direct values in parametrizations isn't possible with substantial additions / modifications to the existing code, which I want to avoid, so a file would be much better for some test. Without a test file, all test functions that have some relevance to the multi-wavelength case would need to be duplicated. However, even with a test file, some if/else branches would be required to handle parts that conflict with multiple wavelengths in the data, and to get around some vendor-specific stuff (such as many tests using nirx I/O), so it's not as simple as just adding the file name in addition to additional parametrizations.

larsoner commented:
> However, even with a test file, some if/else branches would be required to handle parts that conflict with multiple wavelengths in the data, and to get around some vendor-specific stuff (such as many tests using nirx I/O), so it's not as simple as just adding the file name in addition to additional parametrizations.

Yeah that's okay -- I'd expect some new branches to be needed. And ideally some new loops that generalize the 2-wavelength to the 2+-wavelength case would be great where possible, rather than branching. But I imagine some will be unavoidable.

> All in all, I think adding a test file to the other repo, and using it in test_snirf.py (instead of generating a .snirf file in a fixture), and in test_nirs.py might be the best next step. As for the other tests (beer-lambert, od, sci), keeping a fixture might be OK. That fixture would either generate the data manually as now, or read in the .snrif file instead.

Sure feel free to add one of your generated ones to `mne-testing-data` and I'll cut a testing version release and you can use it in this PR

zEdS15B3GCwq commented:
Pushed a new version of the tests. In the end, I decided to base all tests on a sample file, otherwise it would've been difficult to integrate the multi-wavelength data into existing tests.

I'll open a PR to the testing data repo next, the current tests will keep failing until that's done.

Some explanation of what I did:

- `mne/io/snirf/tests/test_snirf.py`: basic import test uses labnirs file, added a test with hard-coded values to test that the labnirs file was correctly loaded (in hindsight, perhaps I should've named this test differently, including labnirs in the name)
- `mne/preprocessing/nirs/tests/test_optical_density.py`: `test_optical_density()` now also read the labnirs file; refactored the test a bit, and added a check to verify that channel names are unchanged.
- `mne/preprocessing/nirs/tests/test_scalp_coupling_index.py`: `test_scalp_coupling_index_multi_wavelength` added that uses the sample labnirs file and also tests cases specific to multi-wavelength files. It's based on `test_scalp_coupling_index()` so there's some overlap. Perhaps the two could be unified...
- `mne/preprocessing/nirs/tests/test_beer_lambert_law.py`: `test_beer_lambert` now includes the labnirs file, if/else branches take care of nirx, fif and snirf imports. Added checks to ensure that the number of channels are reduced to 2 per S-D pair by the BLL step, and that channel prefixes are preserved. Test `test_beer_lambert_unordered_errors()` was moved to `test_nirs.py` as all the checks therein actually depended on `_check_channels_ordered()` and not BLL.
- `mne/preprocessing/nirs/tests/test_nirs.py` has extensive changes: I included the labnirs file in most general tests that aren't specific to a file type or dataset, or where I thought adding multi-wavelength data would be valuable. I took the logic from tests from `test_fnirs_channel_naming_and_order_readers` and `test_beer_lambert_unordered_errors` (BLL test file), and divided them into 2 new tests that are better separated according to what they test: `test_fnirs_channel_frequency_ordering` and `test_nirs_channel_grouping`. All the changed tests were adapted to also work with multi-wavelength data.

larsoner commented on mne/preprocessing/nirs/tests/test_scalp_coupling_index.py:
```diff
@@ -76,3 +79,84 @@ def test_scalp_coupling_index(fname, fmt, tmp_path):
+    times = np.arange(raw.n_times) / raw.info["sfreq"]
+    signal = np.sin(2 * np.pi * 1.0 * times) + 1
+    assert len(raw.ch_names) >= 15 * 3
+    rng = np.random.default_rng()
```
Better to seed, otherwise we hit unreproducible, random errors in CIs etc.
```suggestion
    rng = np.random.default_rng(3289745)
```

zEdS15B3GCwq replied:
Sounds reasonable.

larsoner reviewed (commented):
Otherwise I think it looks good! Going to push (and already pushed a little doc commit), then I think we can probably merge
